horovod/tensorflow/__init__.py

```diff
@@ -401,15 +432,15 @@ def __init__(self, optimizer, name=None, use_locking=False, device_dense='',
                     device_sparse='', compression=Compression.none,
                     sparse_as_dense=False, op=Average, gradient_predivide_factor=1.0,
                     backward_passes_per_step=1, average_aggregated_gradients=False,
-                    num_groups=0):
+                    groups=None):
```

Comment:
We shouldn't completely remove the `num_groups` argument in this PR. Instead, assign `num_groups` to `groups` at the beginning of the `__init__` function and print a deprecation warning that `num_groups` will be removed in a future Horovod version and users should use `groups` instead. For example:
```
if num_groups != 0:
    warnings.warn('Parameter `num_groups` has been replaced by `groups` and will be removed in v0.23.0.', DeprecationWarning)
```

Reply:
I have updated the code for the public APIs. Shall I retain the `num_groups` in constructor of inner classes like `_DistributedOptimizer`?

Reply:
We should remove `num_groups` from inner classes at this point. You may need to modify some files in the TF Keras implementation, like this here: https://github.com/horovod/horovod/blob/2a775b2e059817093e476a5a44433426dda35ae5/horovod/_keras/__init__.py#L40-L48
I realize that the grouping functionality wasn't actually added to the Keras `DistributedOptimizer` constructor, but you don't need to worry about that here. We can fix that in a follow up PR.

Reply:
@romerojosh 
I just tried to add the support for keras. As for mxnet, I have no experience with it, so there is nothing I can do...

BTW, I've also made some changes in `LocalGradientAggregationHelper` and `LocalGradientAggregationHelperEager`, because we'll need variable lists as inputs to `allreduce_grads`. I wonder if there is any unitest I can add for the new functionality, because for now I am just running code in the example folder with different settings and that is not comprehensive enough...

Reply:
@zhuzilin No worries on MXNet and thanks for covering TF and PyTorch. Indeed, there is a big gap in unit testing the `DistributedOptimizer`, especially the with any optional arguments. We have an issue (#2132) open for it, but we still haven't added more testing. Your changes here look good to me so this is not a blocker for me.

@tgaddair I think we should make an effort to resolve #2132 and get more extensive tests put together before the next major release.